### ISSUE TITLE
Fix `indentation` false negatives for `postcss-styled-syntax` custom syntax

### ISSUE THREAD

**Issue by rishitank** [stylelint-stylistic/stylelint-stylistic#39]
### What minimal example or steps are needed to reproduce the bug?

For example, with code such as this:
```
const StyledContainer = styled.div`
  background: #FFFF;
`;
```

### What minimal configuration is needed to reproduce the bug?
```
module.exports = {
  customSyntax: 'postcss-styled-syntax',
  extends: [
    'stylelint-config-recommended',
    '@stylistic/stylelint-config'
  ],
  ignoreFiles: [
    '**/*.scss'
  ],
  plugins: [
    '@stylistic/stylelint-plugin'
  ],
  rules: {
    '@stylistic/indentation': 2
  }
}
```

### How did you run Stylelint?

```
npx stylelint --fix
```

### Which Stylelint-related dependencies are you using?

```
{
    "@stylistic/stylelint-config": "^2.0.0",
    "@stylistic/stylelint-plugin": "^3.0.1",
    "postcss-styled-syntax": "^0.6.4",
    "stylelint": "^16.9.0",
    "stylelint-config-recommended": "^14.0.1"
}
```

### What did you expect to happen?

That no errors would be reported

### What actually happened?

I get the error:
```
Stylelint: Expected indentation of 0 spaces (@stylistic/indentation)
```

### Do you have a proposal to fix the bug?

Support CSS in JS where there should be an initial indentation.

**Comment by firefoxic** [comment]
@rishitank Thanks for the report! I created a minimal reproducible example in the [demo](https://stylelint.io/demo/#N4Igxg9gJgpiBc4IDsDOAXABAZXQTwBsYoBhFdAQwEtkYAnTAXkw0OIDooqA3AAwB1kmTACMKYANYBzOhACuyKPEwBiAGIa1AbkG8tIADQgAZlSIA5CgFs4iGAA9rAByLsAVqkNJkpqQhDAgsL84HIYEFbYeMiU9iHKIU4QGGCoqAC0rERQmdGxIQZBmCEO6DCKqPGYANpFwSAAAllUGFRgAPRZMAQ06OmQPlRSIUUAuoVCxSB0ckSVCJiBk-VN+D2tHTSwMRToVChVAExFAL6CZ8heA75qEHRWu-4eKF6wTp6IS8IrzRuda91ev0UL4qiECLsYBgCnUpkkUmlMgCcqg8hQ4gtwZDoYZYSEuj0YmCQBCyjjziATkA).

I've labeled the issue as ready to implement. Please consider [contributing](https://github.com/stylelint-stylistic/stylelint-stylistic/blob/main/CONTRIBUTING.md) if you have time.

There are [steps on how to fix a bug in a rule](https://github.com/stylelint-stylistic/stylelint-stylistic/blob/main/docs/developer-guide/rules.md#fix-a-bug-in-a-rule) in the Developer guide.

**Comment by MorevM** [comment]
I'll take this, it'll be ready in a week.

**Comment by MorevM** [comment]
After some research, I've come to the conclusion that it's very hard to do it well (I understand the pain from this [issue](https://github.com/typescript-eslint/typescript-eslint/issues/1824), though prettier is still terrible).

The problem is that with a mix of JS and CSS, it becomes difficult to define a base level of indentation.

For simple cases like this, we can inherit the indentation level from the line where the expression starts:

```ts
const StyledContainer = styled.div`
  background: #FFFF;
`;
```

The indentation of `const StyledContainer = styled.div` line is `0`, so the indentation of CSS content should be `1`.
It's ok. 

---

**The first problem:** indentation of JS line itself.
If its indentation is incorrect (for example, 3 spaces instead of two), the calculation of the required indentation for CSS will be incorrect.
JS string indentation can be controlled via <https://eslint.style/rules/default/indent>, this is outside the scope of this plugin. 
Just noting that if it is not done, it will not work correctly.

---

**The second problem:** indentation of inner blocks due to JS variability.

Lets say this codeblock:
```ts
const Component = styled.p`
  ${(props) =>
    props.isPrimary
      ? css`
        background: green;
        `
     // `else` condition omitted for readability
  }
`;
```

Probably you want this to be
```ts
      ? css`
          background: green;
        `
```

But this will be

```ts
      ? css`
        background: green;
        `
```

because the line `? css` indentation level is 3, the CSS string indentation should be 4.
It becomes even harder if we operate with tabs, not spaces.

---

I plan to make support for simple cases using the rule: “plus one level to the indentation of the string where the expression starts”. 
I do not plan to align the closing character of tagged function (this refers to JS code).

If you guys have additional scenarios to consider or additional ideas on how to make it better, you are welcome to comment now - I'm going to look into it this weekend.

**Comment by MorevM** [comment]
@rishitank and other interested

[The change](https://github.com/stylelint-stylistic/stylelint-stylistic/pull/41) is ready for review,

I don't really use styled components in my work, so I might have missed some usage scenarios.
Can you look at the tests and tell me if everything is covered?

**Comment by rishitank** [comment]
> @rishitank and other interested
> 
> [The change](https://github.com/stylelint-stylistic/stylelint-stylistic/pull/41) is ready for review,
> 
> I don't really use styled components in my work, so I might have missed some usage scenarios. Can you look at the tests and tell me if everything is covered?
> 
> [Online demo](https://stylelint.io/demo/#N4Igxg9gJgpiBc4IDsDOAXABAZXQTwBsYoBhFdAQwEtkYAnTAXkw0OIDooqA3AAwB1kmTACMKYANYBzOhACuyKPEwBiAGIa1AbkG8tIADQgAZlSIA5CgFs4iGAA9rAByLsAVqkNJkpqQhDAgsL84HIYEFbYeMiU9iHKIU4QGGCoqAC0rERQmdGxIQZBmIkEclI0qPGYANpFwSAAAllUGFRgAPRZMAQ06OkuZTQhRQC6hULFIHRyRJUImIET9U34Pa0dNLAxFOhUKFUATEUAvoKnyF6QPlRSahB0Vjv+HihesE6eiIvCy83rnatur1+qVysgqiEABbodAfeDtdqQWAECAUKDscroSFyETsSBWAFsHoxTKrFq7DpdYl9P4U9qUOgYgBe7ToMGMqHakJgaM5pns7SSKTSpLYOVQeQo9nSmxg212+0MdUmQvQqQyXXFkri8xCBB2MAwBWVISpvQhIH16EN6GGyGOIGOQA).

Thank you for this fix, I believe the test scenarios are covered. Regarding nested css blocks, we usually get around this by assigning it to a separate variable to improve readability and maintainability, and so that ESLint indent would kick in in those scenarios instead.

**Comment by MorevM** [comment]
@rishitank 

The fix is included in the just released version 3.1, update the package to test on your codebase.
Feel free to reopen if something is still not working as it should. :)

**Comment by rishitank** [comment]
Hi @MorevM, I have had a chance to play with this and found that it can conflict with ESLint indentation rules when using JS expressions inside the styled component. For example:
```
const Option = styled.div`
  &:hover {
    background-color: ${({ colours, isOptionSelected }) => (
      isOptionSelected
        ? colours.backgroundColour
        : colours.backgroundColourOnHover
    )};
  }
`;
```

However this conflicts with Stylelint's indentation rule of 2 spaces:
```
const Option = styled.div`
  &:hover {
    background-color: ${({ colours, isOptionSelected }) => (
        isOptionSelected
        ? colours.backgroundColour
        : colours.backgroundColourOnHover
    )};
  }
`;
```

And it seems both indent rules are fighting for control. Ideally it would be good if I could disable ESLint indent rules for nested JS expressions inside styled components, but I am not sure if that is possible.

**Comment by MorevM** [comment]
Hi @rishitank 

Sorry for the delayed response.
That's exactly what I meant by “indentation of inner blocks due to JS variability” :(

I'm willing to give another chance for a fix, but in general it's always going to be like “add another crutch when someone else has done something else using JS”

Hopefully I'll be able to do something about it and that'll be the end of it.
I plan to remove indentation checking for property values if the styled components syntax is used and the property value starts with `${` - I think that covers this. As long as someone doesn't request dynamic properties using JS expressions :)

I plan to deal with it today/tomorrow.

**Comment by rishitank** [comment]
Thank you for looking into this, is there a planned release for this fix?

**Comment by MorevM** [comment]
@rishitank there is 3.1.1 release with this fix included :)
